keps/sig-api-machinery/NNNN-watch-list/README.md

```diff
@@ -0,0 +1,956 @@
+
+Step 2a: Where does the initial data come from?
+
+During construction, the cacher creates the reflector and the watchCache.
+Since the watchCache implements the Store interface it is used by the reflector to store all data it has received from etcd.
+Note: Right now the reflector used to populate the watchCache uses the old semantics of LIST/WATCH.
+This could be optimized in the future (but out-of-scope here) by establishing a WATCH directly with etcd and waiting for the bookmarkAfterResourceVersion.
+
+Step 2b: What happens when new events are received while the cacheWatcher is sending initial data?
+
+The cacher maintains a list of all current watchers (cacheWatcher) and a separate goroutine (dispatchEvents) for delivering new events to the watchers.
+New events are added via the cacheWatcher.nonblockingAdd method that adds an event to the cacheWatcher.incoming channel.
+The cacheWatcher.incoming is a buffered channel and has a different size for different Resources (10 or 1000).
+Since the cacheWatcher starts processing the cacheWatcher.incoming channel only after sending all initial events it might block once its buffered channel tips over.
+In that case, it will be added to the list of blockedWatchers and will be given another chance to deliver an event after all nonblocking watchers have sent the event.
+All watchers that have failed to deliver the event will be closed.
+
+Note: Given the above, there is a possibility of sending initial data taking too long and failing the given time budget. 
+In that case, the cacheWatcher will be closed and the client (reflector) would have to reconnect.
+
+Note 2: Currently, there is a timeout for LIST requests of 60s.
+That means a slow reflector might fail synchronization as well and would have to re-establish the connection.
+
+Step 2c: How bookmarks are delivered to the cacheWatcher?
+
+First of all, the primary purpose of bookmark events is to deliver the current resourceVersion to watchers, continuously even without regular events happening.
+There are two sources of resourceVersions.
+The first one is regular events that contain RVs besides objects.
+The second one is a special type of etcd event called progressNotification delivering the most up-to-date revision with the given interval.
+As already mentioned in 2a the watchCache is driven by the reflector.
+Every event will be eventually propagated from the watchCache to the cacher.processEvent method.
+For simplicity, we can assume that the processEvent method will simply update the resourceVersion maintained by the cacher.
+
+At regular intervals, the cacher checks expired watchers and tries to deliver a bookmark event.
+As of today, the interval is set to 1 second.
+The bookmark event contains an empty object and the current resourceVersion.
+By default, a watchCache expires roughly every 1 minute.
+
+The expiry interval initially will be decreased to 1 second in this feature's code-path.
```

Comment:
This won't scale. Delivering a bookmark every 1s to every watcher will blow us up (unless I misunderstood what you want to do here).

Reply:
I think that the default delivery (expiry) time for bookmarks events is 1 minute. That means an initial request (RV=0 and the resourceVersionMatch set) would have to wait up to 60s. I propose to decrease that time only for watchers in that special state.